### DISCUSSION TITLE
Using Shap in case of data when all features are not applicable

### DISCUSSION THREAD

**Discussion by datascienceenthu1** [shap/shap#3109] [Q&A]
Suppose we train a single binary classification model on 3 cohorts C1, C2, C3. Suppose feature F1 is available only in C1, is super predictive but is not available or populated at all in C2 and C3. (In the model for F1- let's say we give a default value  for Cohorts C2 and C3 )
When we calculate Shap value of features for a record from Cohort C2, we get a Shap value for F1.  
1.	Is there a way to prevent feature F1 from being considered at all for data from C2/C3 ? 
2.	If not how do we rationalize this ? The Shap value for a record belonging to C2 would be arrived at by testing multiple possible values for F1, which in reality would not be taken at all in C2 and C3.

**Comment by connortann** [comment]
I think the question here is more related to building a model that generalises well between cohorts, than the use of Shapley values to explain the model.

> Is there a way to prevent feature F1 from being considered at all for data from C2/C3 ?

The model you describe uses imputed values of F1 in examples in C2/C3. You could try handling missing values in a different way, perhaps by avoid imputing and using a model that can handle missing values natively (e.g. a decision tree). This model would be less likely to be led astray by fictional imputed values.

Another approach could be to build a separate model that does not use the feature F1, for use in contexts where F1 is not available.

> If not how do we rationalize this ? 

If the Shapley value for a constant imputed feature is nonzero, one interpretation could be that the feature is acting as an indicator that the example is probably a member of a cohort in which the feature is typically missing, which in itself could be informative.

I'd argue that the _absence_ of a feature can still be informative for a model. As an example, consider a model that predicts a rare disease from blood tests. A feature could have three values: [test positive, test negative, blood test not taken]. Suppose Cohort 3 has a low incidence of the disease, and no blood tests were performed in this cohort. A value of "blood test not taken" is informative: it suggests the individual is probably in Cohort 3, and hence probably doesn't have the disease.

Hope this helps, let me know if this answers your question or not!
